// @rpldy/shared-ui/lib/cjs/eventListenerHooks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useAllAbortListener = exports.useRequestPreSend = exports.useItemFinalizeListener = exports.useItemAbortListener = exports.useItemErrorListener = exports.useItemCancelListener = exports.useItemProgressListener = exports.useItemFinishListener = exports.useItemStartListener = exports.useBatchProgressListener = exports.useBatchAbortListener = exports.useBatchCancelledListener = exports.useBatchFinishListener = exports.useBatchStartListener = exports.useBatchAddListener = void 0;

var _uploader = require("@rpldy/uploader");

var _utils = require("./utils");

const useBatchAddListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.BATCH_ADD, false);
exports.useBatchAddListener = useBatchAddListener;
const useBatchStartListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.BATCH_START);
exports.useBatchStartListener = useBatchStartListener;
const useBatchFinishListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.BATCH_FINISH);
exports.useBatchFinishListener = useBatchFinishListener;
const useBatchCancelledListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.BATCH_CANCEL);
exports.useBatchCancelledListener = useBatchCancelledListener;
const useBatchAbortListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.BATCH_ABORT);
exports.useBatchAbortListener = useBatchAbortListener;
const useBatchProgressListener = (0, _utils.generateUploaderEventHookWithState)(_uploader.UPLOADER_EVENTS.BATCH_PROGRESS, batch => ({ ...batch
}));
exports.useBatchProgressListener = useBatchProgressListener;
const useItemStartListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.ITEM_START);
exports.useItemStartListener = useItemStartListener;
const useItemFinishListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.ITEM_FINISH);
exports.useItemFinishListener = useItemFinishListener;
const useItemCancelListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.ITEM_CANCEL);
exports.useItemCancelListener = useItemCancelListener;
const useItemErrorListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.ITEM_ERROR);
exports.useItemErrorListener = useItemErrorListener;
const useItemAbortListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.ITEM_ABORT);
exports.useItemAbortListener = useItemAbortListener;
const useItemFinalizeListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.ITEM_FINALIZE);
exports.useItemFinalizeListener = useItemFinalizeListener;
const useItemProgressListener = (0, _utils.generateUploaderEventHookWithState)(_uploader.UPLOADER_EVENTS.ITEM_PROGRESS, item => ({ ...item
}));
exports.useItemProgressListener = useItemProgressListener;
const useRequestPreSend = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.REQUEST_PRE_SEND, false);
exports.useRequestPreSend = useRequestPreSend;
const useAllAbortListener = (0, _utils.generateUploaderEventHook)(_uploader.UPLOADER_EVENTS.ALL_ABORT, false);
exports.useAllAbortListener = useAllAbortListener;
