// @rpldy/uploader/lib/cjs/processor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _queue = _interopRequireDefault(require("./queue"));

var _batchItemsSender = _interopRequireDefault(require("./batchItemsSender"));

var _batch = _interopRequireDefault(require("./batch"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = (trigger, cancellable, options, uploaderId) => {
  const sender = (0, _batchItemsSender.default)(),
        queue = (0, _queue.default)(options, trigger, cancellable, sender, uploaderId);
  const runCancellable = queue.runCancellable;
  return {
    process: (batch, batchOptions) => {
      queue.uploadBatch(batch, batchOptions);
    },
    abortBatch: batchId => {
      queue.abortBatch(batchId);
    },
    abort: id => {
      if (id) {
        queue.abortItem(id);
      } else {
        queue.abortAll();
      }
    },
    addNewBatch: (files, uploaderId, processOptions) => {
      const batch = (0, _batch.default)(files, uploaderId, processOptions);
      return queue.addBatch(batch, processOptions);
    },
    runCancellable,
    clearPendingBatches: () => {
      queue.clearPendingBatches();
    },
    processPendingBatches: uploadOptions => {
      queue.uploadPendingBatches(uploadOptions);
    }
  };
};

exports.default = _default;
