// @rpldy/uploader/lib/cjs/queue/batchHelpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getIsBatchFinalized = exports.incrementBatchFinishedCounter = exports.removePendingBatches = exports.preparePendingForUpload = exports.detachRecycledFromPreviousBatch = exports.getBatchFromState = exports.getIsItemBatchReady = exports.triggerUploaderBatchEvent = exports.cleanUpFinishedBatches = exports.getBatchDataFromItemId = exports.isItemBelongsToBatch = exports.getBatchFromItemId = exports.cancelBatchForItem = exports.isNewBatchStarting = exports.loadNewBatchForItem = void 0;

var _shared = require("@rpldy/shared");

var _simpleState = require("@rpldy/simple-state");

var _consts = require("../consts");

const BATCH_READY_STATES = [_shared.BATCH_STATES.ADDED, _shared.BATCH_STATES.PROCESSING, _shared.BATCH_STATES.UPLOADING];
const BATCH_FINISHED_STATES = [_shared.BATCH_STATES.ABORTED, _shared.BATCH_STATES.CANCELLED, _shared.BATCH_STATES.FINISHED];

const getBatchFromState = (state, id) => state.batches[id].batch;

exports.getBatchFromState = getBatchFromState;

const getBatch = (queue, id) => {
  return getBatchFromState(queue.getState(), id);
};

const isItemBelongsToBatch = (queue, itemId, batchId) => {
  return queue.getState().items[itemId].batchId === batchId;
};

exports.isItemBelongsToBatch = isItemBelongsToBatch;

const getBatchDataFromItemId = (queue, itemId) => {
  const state = queue.getState();
  const item = state.items[itemId];
  return state.batches[item.batchId];
};

exports.getBatchDataFromItemId = getBatchDataFromItemId;

const getBatchFromItemId = (queue, itemId) => {
  return getBatchDataFromItemId(queue, itemId).batch;
};

exports.getBatchFromItemId = getBatchFromItemId;

const removeBatchItems = (queue, batchId) => {
  const batch = getBatch(queue, batchId);
  queue.updateState(state => {
    batch.items.forEach(({
      id
    }) => {
      delete state.items[id];
      const index = state.itemQueue.indexOf(id);

      if (~index) {
        state.itemQueue.splice(index, 1);
      }
    });
  });
};

const removeBatch = (queue, batchId) => {
  queue.updateState(state => {
    delete state.batches[batchId];
  });
};

const cancelBatchForItem = (queue, itemId) => {
  const batch = getBatchFromItemId(queue, itemId),
        batchId = batch.id;

  _shared.logger.debugLog("uploady.uploader.batchHelpers: cancelling batch: ", {
    batch
  });

  queue.updateState(state => {
    const batch = getBatchFromState(state, batchId);
    batch.state = _shared.BATCH_STATES.CANCELLED;
  });
  triggerUploaderBatchEvent(queue, batchId, _consts.UPLOADER_EVENTS.BATCH_CANCEL);
  removeBatchItems(queue, batchId);
  removeBatch(queue, batchId);
};

exports.cancelBatchForItem = cancelBatchForItem;

const isNewBatchStarting = (queue, itemId) => {
  const batch = getBatchFromItemId(queue, itemId);
  return queue.getState().currentBatch !== batch.id;
};

exports.isNewBatchStarting = isNewBatchStarting;

const loadNewBatchForItem = (queue, itemId) => {
  const batch = getBatchFromItemId(queue, itemId);
  return queue.runCancellable(_consts.UPLOADER_EVENTS.BATCH_START, batch).then(isCancelled => {
    if (!isCancelled) {
      queue.updateState(state => {
        state.currentBatch = batch.id;
      });
    }

    return !isCancelled;
  });
};

exports.loadNewBatchForItem = loadNewBatchForItem;

const cleanUpFinishedBatches = queue => {
  //TODO: schedule clean up on requestAnimationFrame
  const state = queue.getState();
  Object.keys(state.batches).forEach(batchId => {
    const {
      batch,
      finishedCounter
    } = state.batches[batchId];
    const {
      orgItemCount
    } = batch; //shouldnt be the case, but if wasnt cleaned before, it will now

    const alreadyFinalized = getIsBatchFinalized(batch);

    if (orgItemCount === finishedCounter) {
      queue.updateState(state => {
        const batch = getBatchFromState(state, batchId); //set batch state to FINISHED before triggering event and removing it from queue

        batch.state = alreadyFinalized ? batch.state : _shared.BATCH_STATES.FINISHED;

        if (state.currentBatch === batchId) {
          state.currentBatch = null;
        }
      });

      _shared.logger.debugLog(`uploady.uploader.batchHelpers: cleaning up batch: ${batch.id}`);

      if (!alreadyFinalized) {
        triggerUploaderBatchEvent(queue, batchId, _consts.UPLOADER_EVENTS.BATCH_FINISH);
      }

      removeBatchItems(queue, batchId);
      removeBatch(queue, batchId);
    }
  });
};

exports.cleanUpFinishedBatches = cleanUpFinishedBatches;

const triggerUploaderBatchEvent = (queue, batchId, event) => {
  const state = queue.getState(),
        batch = getBatchFromState(state, batchId),
        //get the most uptodate batch data
  stateItems = state.items;
  const eventBatch = { ...(0, _simpleState.unwrap)(batch),
    items: batch.items.map(({
      id
    }) => (0, _simpleState.unwrap)(stateItems[id]))
  };
  queue.trigger(event, eventBatch);
};

exports.triggerUploaderBatchEvent = triggerUploaderBatchEvent;

const getIsItemBatchReady = (queue, itemId) => {
  const batch = getBatchFromItemId(queue, itemId);
  return BATCH_READY_STATES.includes(batch.state);
};

exports.getIsItemBatchReady = getIsItemBatchReady;

const detachRecycledFromPreviousBatch = (queue, item) => {
  const {
    previousBatch
  } = item;

  if (item.recycled && previousBatch && queue.getState().batches[previousBatch]) {
    const {
      id: batchId
    } = getBatchFromItemId(queue, item.id);

    if (batchId === previousBatch) {
      queue.updateState(state => {
        const batch = getBatchFromState(state, batchId);
        const index = batch.items.findIndex(({
          id
        }) => id === item.id);

        if (~index) {
          batch.items.splice(index, 1);
        }
      });
    }
  }
};

exports.detachRecycledFromPreviousBatch = detachRecycledFromPreviousBatch;

const preparePendingForUpload = (queue, uploadOptions) => {
  queue.updateState(state => {
    //remove pending state from pending batches
    Object.keys(state.batches).forEach(batchId => {
      const batchData = state.batches[batchId];
      const {
        batch,
        batchOptions
      } = batchData;

      if (batch.state === _shared.BATCH_STATES.PENDING) {
        batch.items.forEach(item => {
          item.state = _shared.FILE_STATES.ADDED;
        });
        batch.state = _shared.BATCH_STATES.ADDED;
        batchData.batchOptions = (0, _shared.merge)({}, batchOptions, uploadOptions);
      }
    });
  });
};

exports.preparePendingForUpload = preparePendingForUpload;

const removePendingBatches = queue => {
  const batches = queue.getState().batches;
  Object.keys(batches).filter(batchId => batches[batchId].batch.state === _shared.BATCH_STATES.PENDING).forEach(batchId => {
    removeBatchItems(queue, batchId);
    removeBatch(queue, batchId);
  });
};

exports.removePendingBatches = removePendingBatches;

const incrementBatchFinishedCounter = (queue, batchId) => {
  queue.updateState(state => {
    state.batches[batchId].finishedCounter += 1;
  });
};

exports.incrementBatchFinishedCounter = incrementBatchFinishedCounter;

const getIsBatchFinalized = batch => BATCH_FINISHED_STATES.includes(batch.state);

exports.getIsBatchFinalized = getIsBatchFinalized;
